mlton/codegen/llvm-codegen/llvm-codegen.fun

```diff
@@ -266,17 +266,88 @@ fun mkgep (lhs, ty, arg, idcs) =
         concat ["\t", lhs, " = getelementptr inbounds ", getTypeFromPointer ty, ", ", ty, " ", arg, ", ", indices, "\n"]
     end
 
+
+
+structure Metadata = struct
+   datatype t =
+      Unnamed of int
+   fun str (Unnamed i) = "!" ^ (Int.toString i)
+
+   val metaDataCounter = ref 0
+   fun new () =
+      let
+         val i = !metaDataCounter
+         val () = Int.inc metaDataCounter
+      in
+         Unnamed i
+      end
+   fun reset () =
+      metaDataCounter := 0
+
+   val rec addSep =
+      fn [] => ""
+       | [s] => s
+       | s :: ss' => concat [s, ", ", addSep ss']
+
+   fun defineNode (t, ts) =
+      concat
+         [str t,
+          " = !{",
+          (addSep o List.map) (ts, str),
+          "}"]
+end
+
+structure SimpleOper = struct
+
+   datatype t = Stack
+              | Offset
+              | SequenceOffset
+              | Other
+   val equals : t * t -> bool = op =
+   val hash =
+      fn Stack => Hash.permute 0w0
+       | Offset => Hash.permute 0w1
+       | SequenceOffset => Hash.permute 0w2
+       | Other => Hash.permute 0w3
+   val fromOper =
+      fn Operand.StackOffset _ => Stack
+       | Operand.Offset _ => Offset
+       | Operand.SequenceOffset _ => SequenceOffset
```

Comment:
These aren't quite accurate.  For `Operand.Offset` and `Operand.SequenceOffset`, we need to look at the type of the `base` field, because both operands can be used for non-ML data (e.g., `Operand.Offset` is used to access fields of `GCState` and Operand.SequenceOffset` is used to implement the `CPointer_{get,set}<ty>` primitives).

Reply:
I've filtered based on types and put non-objptrs into the Other category (which is disjoint, rather than unknown). That should be sound for our flows, as far as I can tell.